plugin/src/rules/no_useless_escape.rs

```diff
@@ -0,0 +1,2239 @@
+        }
+    } else {
+        is_quote_escape = escaped_char == node.text(context).chars().next().unwrap();
+    }
+
+    if is_unnecessary_escape && !is_quote_escape {
+        report(
+            node,
+            template_chunk_and_start
+                .map(|template_chunk_and_start| template_chunk_and_start.1)
+                .unwrap_or_else(|| node.start_byte())
+                + captures.get(0).unwrap().start(),
+            captures[0][1..].to_owned(),
+            context,
+        );
+    }
+}
+
+fn check<'a>(
+    node: Node<'a>,
+    template_chunk_and_start: Option<(Cow<'a, str>, usize)>,
+    context: &QueryMatchContext<'a, '_>,
+) {
+    let is_template_element = template_chunk_and_start.is_some();
+
+    if is_template_element
+        && node.parent().matches(|parent| {
+            is_tagged_template_expression(parent) && parent.field("arguments") == node
+        })
+    {
+        return;
+    }
+
+    if matches!(node.parent().unwrap().kind(), JsxAttribute) {
+        return;
+    }
+
+    let value = template_chunk_and_start
+        .as_ref()
+        .map(|template_chunk_and_start| template_chunk_and_start.0.clone())
+        .unwrap_or_else(|| node.text(context));
+
+    for captures in regex!(r#"\\[^\d]"#).captures_iter(&value) {
+        validate_string(node, captures, template_chunk_and_start.as_ref(), context);
+    }
+}
+
+pub fn no_useless_escape_rule() -> Arc<dyn Rule> {
+    rule! {
+        name => "no-useless-escape",
+        languages => [Javascript],
+        messages => [
+            unnecessary_escape => "Unnecessary escape character: \\{{character}}.",
+            remove_escape => "Remove the `\\`. This maintains the current functionality.",
+            remove_escape_do_not_keep_semantics => "Remove the `\\` if it was inserted by mistake.",
+            escape_backslash => "Replace the `\\` with `\\\\` to include the actual backslash character.",
+        ],
+        listeners => [
+            r#"
+              (template_string) @c
+            "# => |node, context| {
+                for (chunk, chunk_start) in get_template_string_chunks(node, context) {
+                    check(node, Some((chunk, chunk_start)), context);
+                }
+            },
+            r#"
+              (string) @c
+            "# => |node, context| {
+                check(node, None, context);
+            },
+            r#"
+              (regex) @c
+            "# => |node, context| {
+                let pattern = node.field("pattern").text(context);
+                let flags = node.child_by_field_name("flags").map(|flags| flags.text(context));
+                let unicode = flags.as_ref().matches(|flags| flags.contains('u'));
+                let unicode_sets = flags.as_ref().matches(|flags| flags.contains('v'));
+
+                let arena = AllArenas::default();
+                let mut parser = RegExpParser::new(&arena, None);
+                let pattern_as_wtf16: Wtf16 = (&*pattern).into();
```

Comment:
This strikes me as potentially expensive enough that it would be worth eg scanning the regex pattern text for a backslash before doing this more "heavyweight" checking?